docs/testing-strategy.md

```diff
@@ -0,0 +1,154 @@
+a machine perspective. They are used to determine that for a given input and
+application state a particular response is returned. These differ from feature
+tests in that they test a single endpoint at a time and can exhaust the logical
+outcomes of that endpoint where some situations may be difficult for
+a user to fall into.
+
+Characteristics of request tests:
+
+- they should test the variety of responses an endpoint returns, checking
+  aspects of the response such as status code and any flash messages;
+- they shouldn't test responses where there isn't specific logic written for
+  the scenario;
+- they should assert against the HTTP responses involved in the request, and
+  not side-effects such as database changes, these are better suited to [unit
+  testing](#unit-tests) an [interactor][interactors];
+- they should focus on the effects of a single endpoint with a single HTTP
+  method at a time, for example `POST /documents`, for multiple endpoints
+  consider a [feature test](#feature-tests);
+- they act as integration tests and thus shouldn't mock application code;
+- they can mock external API calls, but these should not be asserted against -
+  this is better suited to [unit tests](#unit-tests);
+- they shouldn't be used to extensively test HTML output, this is better suited
+  by [view tests](#view-tests);
+- they may not be necessary for places where a response doesn't have logic
+  and already has coverage provided by a feature test.
+
+## View tests
+
+The purpose of view tests is to assert that the expected HTML is rendered in a
+particular scenario. Typically it isn't necessary to have view tests as the
+use of complex logic is discouraged in views and often aspects of views are
+implicitly tested in feature or request tests. However when it is suited for
+us to test particular HTML output they are the most appropriate choice.
+
+Characteristics of view tests:
+
+- they should make assertions based on logic in the view and not be used to
+  assert that HTML is as we wrote it;
+- inputs into tests and dependent objects can be mocks of the expected object;
+- they should follow the conventions of
+[RSpec Rails view tests][rspec-rails-views].
+
+[whitehall-tests]: https://github.com/alphagov/whitehall/blob/099f53e35cdd0ea63a2349be3766c98e65521ce3/docs/testing.md
+[email-alert-api-tests]: https://github.com/alphagov/email-alert-api/pull/300/commits/bf087c28442629bf672be67441fd224f6d746fa5
+[thoughtbot-rails-tests]: https://thoughtbot.com/blog/how-we-test-rails-applications
+[factory-bot]: https://github.com/thoughtbot/factory_bot
+[rails-views]: https://guides.rubyonrails.org/testing.html#testing-views
+[readable-feature-tests]: https://about.futurelearn.com/blog/how-we-write-readable-feature-tests-with-rspec
+[interactors]: adr/0011-moving-business-logic-out-of-controllers.md
+[rspec-rails-views]: https://relishapp.com/rspec/rspec-rails/v/3-9/docs/view-specs/view-spec
```

Comment:
Could you make this list of link texts alphabetical?

Reply:
I'm not aware that we have a convention for the ordering of links in markdown. I've gone with ordering them based on the order they appear in the doc.